cmd/samples/recipes/signalcounter/workflow_test.go

```diff
@@ -0,0 +1,48 @@
+package main
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/cadence/testsuite"
+	"go.uber.org/cadence/workflow"
+)
+
+type UnitTestSuite struct {
+	suite.Suite
+	testsuite.WorkflowTestSuite
+
+	env *testsuite.TestWorkflowEnvironment
+}
+
+func TestUnitTestSuite(t *testing.T) {
+	suite.Run(t, new(UnitTestSuite))
+}
+
+func (s *UnitTestSuite) SetupTest() {
+	s.env = s.NewTestWorkflowEnvironment()
+	s.env.RegisterWorkflow(sampleSignalCounterWorkflow)
+}
+
+func (s *UnitTestSuite) TearDownTest() {
+	s.env.AssertExpectations(s.T())
+}
+
+func (s *UnitTestSuite) Test_SampleSignalCounterWorkflow() {
```

Comment:
tbh I think this is a fairly "exotic" test.  it's relying on the fact that 11 signals _sent unrealistically quickly_ will all arrive, and starve the `s.AddDefault(...)` branch.  this workflow is executing once, and processing an excessive amount of signals (i.e. beyond its limit) due to that.

which is a useful test!  it's also useful that we can demonstrate this behavior in tests, since it can occur in reality (e.g. slow decision task processing, or doing too much work in the loop).  but I don't think it's all that _representative_ of a test for this workflow.  at least as the only test.

that said, I'm not really sure how to test continue-as-new workflows 'realistically'.  it sorta feels like we need two execution modes for tests, one for "execute only once" and one for "execute until all continue-as-news complete".  I think we may have something like that for children workflows, but maybe not for top-level ones...

Reply:
yeah! Also, in the real world the workflow will process several batch of signals in multiple decision task. I don't really know how to test it.